srai/h3.py

```diff
@@ -0,0 +1,166 @@
+"""Utility H3 related functions."""
+
+from typing import Iterable, List, Literal, Tuple, Union, overload
+
+import geopandas as gpd
+import h3
+import numpy as np
+import numpy.typing as npt
+from h3ronpy.arrow.vector import cells_to_wkb_polygons, wkb_to_cells
+from shapely.geometry import Polygon
+from shapely.geometry.base import BaseGeometry
+
+from srai.constants import GEOMETRY_COLUMN, WGS84_CRS
+
+
+def shapely_geometry_to_h3(
+    geometry: Union[BaseGeometry, Iterable[BaseGeometry], gpd.GeoSeries, gpd.GeoDataFrame],
+    h3_resolution: int,
+    buffer: bool = True,
+) -> List[str]:
+    """
+    Convert Shapely geometry to H3 indexes.
+
+    Args:
+        geometry (Union[BaseGeometry, Iterable[BaseGeometry], GeoSeries, GeoDataFrame]):
+            Shapely geometry to be converted.
+        h3_resolution (int): H3 resolution of the cells. See [1] for a full comparison.
+        buffer (bool, optional): Whether to fully cover the geometries with
+            H3 Cells (visible on the borders). Defaults to True.
+
+    Returns:
+        List[str]: List of H3 indexes that cover a given geometry.
+
+    Raises:
+        ValueError: If resolution is not between 0 and 15.
+
+    References:
+        1. https://h3geo.org/docs/core-library/restable/
+    """
+    if not (0 <= h3_resolution <= 15):
+        raise ValueError(f"Resolution {h3_resolution} is not between 0 and 15.")
+
+    wkb = []
+    if isinstance(geometry, gpd.GeoSeries):
+        wkb = geometry.to_wkb()
+    elif isinstance(geometry, gpd.GeoDataFrame):
+        wkb = geometry[GEOMETRY_COLUMN].to_wkb()
+    elif isinstance(geometry, Iterable):
+        wkb = [sub_geometry.wkb for sub_geometry in geometry]
+    else:
+        wkb = [geometry.wkb]
+
+    h3_indexes = wkb_to_cells(
+        wkb, resolution=h3_resolution, all_intersecting=buffer, flatten=True
```

Comment:
I really like the `all_intersecting` parameter name. It seems more descriptive than `buffer` in my opinion. We could replace it in H3 and S2 Regionalizers.

Reply:
neither buffer nor all_intersecting is descriptive enough for me hah. Do we have a better alternative?